Mathlib/CategoryTheory/Closed/FunctorToTypes.lean

```diff
@@ -0,0 +1,135 @@
+morphism `coyoneda.obj (.op c) ⊗ F ⟶ G`. Used to construct a morphism `G ⟶ ihom F H`. -/
+def homEquiv_toFun_app (f : F ⊗ G ⟶ H) (c : C) (y : G.obj c) :
+    (ihom F H).obj c where
+  app := fun d ⟨g, x⟩ ↦ f.app d (x, G.map g.down y)
+  naturality a b h := by
+    ext ⟨g, x⟩
+    have := f.naturality h
+    apply_fun (fun f ↦ f (x, G.map g.down y)) at this
+    dsimp only [coyoneda, uliftFunctor]
+    aesop
+
+@[ext]
+lemma homEquiv_toFun_app_ext {c : C} {f g : (ihom F G).obj c} :
+    f.app = g.app → f = g := NatTrans.ext _ _
+
+/-- Given a morphism `F ⊗ G ⟶ H`, construct a morphism `G ⟶ ihom F H`. -/
+def homEquiv_toFun (f : F ⊗ G ⟶ H) : G ⟶ ihom F H where
+  app := homEquiv_toFun_app f
+  naturality c d g := by
+    ext
+    simp only [ihom, types_comp_apply, coyoneda, uliftFunctor, homEquiv_toFun_app]
+    aesop
+
+@[ext]
+lemma homEquiv_toFun_ext {f g : G ⟶ ihom F H} :
+    f.app = g.app → f = g := NatTrans.ext _ _
+
+/-- Given a morphism `G ⟶ ihom F H`, construct a morphism `F ⊗ G ⟶ H`. -/
+def homEquiv_invFun (f : G ⟶ ihom F H) : F ⊗ G ⟶ H where
+  app c x := (f.app c x.2).app c (Equiv.ulift.symm (𝟙 c), x.1)
+  naturality c d g := by
+    ext ⟨x, y⟩
+    have h := f.naturality g
+    apply_fun (fun f ↦ (f y).app d (Equiv.ulift.symm (𝟙 d), F.map g x)) at h
+    have h' := (f.app c y).naturality g
+    apply_fun (fun f ↦ f (Equiv.ulift.symm (𝟙 c), x)) at h'
+    dsimp only [ihom, coyoneda, uliftFunctor] at h h' ⊢
+    aesop
+
+@[ext]
+lemma homEquiv_invFun_ext {f g : F ⊗ G ⟶ H} :
+    f.app = g.app → f = g := NatTrans.ext _ _
+
+/-- The bijection between morphisms `F ⊗ G ⟶ H` and morphisms `G ⟶ ihom F H`. -/
+def homEquiv (F G H : C ⥤ Type max w v u) : (F ⊗ G ⟶ H) ≃ (G ⟶ ihom F H) where
+  toFun := homEquiv_toFun
+  invFun := homEquiv_invFun
+  left_inv _ := by ext; dsimp only [homEquiv_invFun, homEquiv_toFun, homEquiv_toFun_app]; aesop
+  right_inv f := by
+    ext c y d ⟨g, x⟩
+    have b := f.naturality g.down
+    apply_fun (fun f ↦ (f y).app d (Equiv.ulift.symm (𝟙 d), x)) at b
+    dsimp only [rightAdj, ihom, coyoneda, uliftFunctor] at b ⊢
+    aesop
+
+/-- The adjunction `tensorLeft F ⊣ rightAdj F`. -/
+def adj (F : C ⥤ Type max w v u) : tensorLeft F ⊣ rightAdj F where
+  homEquiv := homEquiv F
+  unit := {
+    app := fun G ↦ homEquiv_toFun (𝟙 _)
+    naturality := fun G H f ↦ by
+      ext c y
+      dsimp [rightAdj, homEquiv_toFun, homEquiv_toFun_app]
+      ext d ⟨g, x⟩
+      dsimp only [Monoidal.tensorObj_obj, comp, Monoidal.whiskerLeft_app, whiskerLeft_apply]
+      rw [Eq.symm (FunctorToTypes.naturality G H f g.down y)]
+  }
+  counit := { app := fun G ↦ homEquiv_invFun (𝟙 _) }
+
+instance closed (F : C ⥤ Type max w v u) : Closed F where
+  rightAdj := rightAdj F
+  adj := adj F
+
+instance cartesianClosed : CartesianClosed (C ⥤ Type max w v u) :=
+  letI _ : MonoidalCategory (C ⥤ Type (max w v u)) := monoidalOfHasFiniteProducts _
```

Comment:
I'm a little confused. We checked earlier that the monoidal structure being used above came from the "chosen" cartesian structure, but this suggests that you're looking at the "noncomputable" version coming from existence of finite products in the sense of `limit`. Can you check that the code above (especially `adj`) is indeed using the chosen cartesian structure?

Reply:
As far as I can tell everything uses the "chosen" cartesian structure. I am checking in the same way as earlier, by seeing if `(F ⊗ G).obj c = ((F.obj c) × (G.obj c))` is true by `rfl`. It seems to be true in the context of `adj`.

Reply:
I see what's happening now. The code you have could be problematic because there are two monoidal instances around. I would just make a `MonoidalClosed` instance instead.

Reply:
Ok, I guess `CartesianClosed` can be worked out later, if at all